File: Dynamic Programming/1388. Pizza With 3n Slices.py
class Solution:
    def maxSizeSlices(self, slices: List[int]) -> int:
        n = len(slices) // 3
        def linear(arr):
            eat = collections.deque([[0] + [-math.inf]*n] * 2) #加0怕有边界问题 因为 i - 1
            res = 0
            for x in arr:
                eat += [i and max(eat[-1][i], eat[-2][i-1]+x) for i in range(n+1)], #1 and 5, return 5
                eat.popleft()
                res = max(res, eat[-1][n])
            return res
        return max(linear(slices[1:]), linear(slices[:-1]))


class Solution:
    def maxSizeSlices(self, slices: List[int]) -> int:
        n = len(slices) // 3 
        def dp(arr):
            eat = collections.deque([[0,]*n]*2)
            for i, x  in enumerate(arr):
                eat += [ i >= 2*j and max(eat[-1][j], eat[-2][j-1] + x) for j in range(n)],
                eat.popleft()
            return eat[-1][-1]
        return max(dp(slices[:-1]), dp(slices[1:]))


class Solution:
    def maxSizeSlices(self, slices: List[int]) -> int:
        n = len(slices) // 3 
        def dp(arr):
            eat = [[0,]*len(arr) for _ in range(2)]
            for i in range(n):
                for j, x in enumerate(arr):
                    if j >= 2*i:
                        eat[i%2][j] = max(eat[i%2][j-1], eat[(i-1)%2][j-2] + x)
            return eat[len(arr)%2][-1]
        return max(dp(slices[:-1]), dp(slices[1:]))

"""

Why do we check j - i + 1 < k * 2 - 1,
instead of something like (k-1) < j - i +1 <= 3 * k

@caohuicn help explain this part:
Because j - i + 1 is not the number of all the slices that are left.
You might have chosen not to take a slice in a previous step,
then there would be leftovers outside of [i, j].
Now if you take i or j (the start or end slice),
one of the slices your friends take will be outside of [i, j],
so the length of [i, j] is reduced by 2, not 3.
Therefore the minimum number is 2 * k - 1
(the last step only requires one slice).

"""

class Solution:
    def maxSizeSlices(self, A):
        @functools.lru_cache(None)
        def dp(i, j, k, cycle=0):
            # i,j = start,end (inclusive)
            # k = remaining
            # cycle (see comment on cycle variable)
            
            if k == 1:
                # one remaining, calculate immediately
                return max(A[i:j + 1])
            
            if j - i + 1 < k * 2 - 1:  # j - i + 1 = 5 有5个数, k = 3， 可以选取index 0, 2, 4, 就不会return -float''inf
                # not possible because not enough elements remain
                return -float('inf')
            
            return max(A[j] +                 # take last element
                       dp(i + cycle, j - 2,   # dp on i to j-2
                          k - 1),             # one less element left to take
                          # (on the cycle variable)
                          # if the last element of the inital array is taken
                          # you cannot take the first element of the inital array
                       # Alternatively, you choose not to take element j
                       dp(i, j - 1,    # dp on i to j-1
                          k))          # same number of elements to take

        return dp(0, len(A) - 1, # dp on 0 to length-1
                  len(A) // 3,   # number of elements to take
                  1)             # see comment on cycle variable
            
        # lru_cache to reduce search space



class Solution:
    def maxSizeSlices(self, slices: List[int]) -> int:
        @functools.lru_cache(None)
        def dp(i, j, k, firstOne = 0):
            if k == 1: 
                return max(slices[i:j+1])
            if j - i + 1 < 2*k - 1 : 
                return float('-inf')
            return max(slices[i] + dp(i+2, j-firstOne, k-1), dp(i+1, j, k))
        return dp(0, len(slices)-1, len(slices)//3, True)

File: Dynamic Programming/1388. Pizza With 3n Slices.cpp
class Solution {
public:
    int maxSizeSlices(vector<int>& slices) {
        int n = (int)slices.size() / 3;
        auto l1 = vector<int>(slices.begin(), slices.end()-1);
        auto l2 = vector<int>(slices.begin()+1, slices.end());
        return max(linear(l1, n), linear(l2, n));
    }
    
private:
    int linear(vector<int>& slices, int n) {
        vector<vector<int>> eat((int)slices.size()+2, vector<int>(n+1, INT_MIN));

        int res = INT_MIN;
        for (int i=0; i<eat.size(); ++i) eat[i][0] = 0;
        for (int i=2; i<eat.size(); ++i) {
            for (int j=1; j<n+1; ++j)
                eat[i][j] = max(eat[i-1][j], eat[i-2][j-1] + slices[i-2]);
            res = max(eat[i][n], res);
        }
        return res;
    }
};

/*
[4,1,2,5,8,3,1,9,7]
stdout

4, 1, 2, 5, 8, 3, 1, 9,
-----------------------
4, 4, 4, 5, 8, 8, 8, 9, 
0, 0, 6, 9, 12, 12, 12, 17, 
0, 0, 0, 0, 14, 14, 14, 21, 

1, 2, 5, 8, 3, 1, 9, 7
---------------------
1, 2, 5, 8, 8, 8, 9, 9, 
0, 0, 6, 10, 10, 10, 17, 17, 
0, 0, 0, 0, 9, 11, 19, 19, 

 

*/

class Solution {
public:
    int maxSizeSlices(vector<int>& slices) {
        vector<int>s1 (slices.begin(), slices.end()-1);
        vector<int>s2 (slices.begin()+1, slices.end());
        int n = slices.size()/3;
        return max(maxSlice(s1, n), maxSlice(s2, n));
    }
    
    int maxSlice(const vector<int>& slices,int n){
        vector<vector<int>>dp(n, vector<int>(slices.size()));
        int res = 0;
        for(int i = 0; i < n; ++i){ //pieces of pizza to pick 
            for(int j = 0; j<slices.size(); ++j){
                if(i == 0)
                    dp[0][j] = max( j >0 ? dp[0][j-1] : 0, slices[j]) ; //只pick 一个pizza
                else if(j >= 2*i) //i = 1, pick 两个pizza, 开始只能拿0, 2
                                //  i = 2, pick 三个pizza, 开始只能pick 0,2,4
                    dp[i][j]  = max(dp[i][j-1], dp[i-1][j-2] + slices[j]);   
                
                res = max(res, dp[i][j]);
            }
        }
        return res;
    }
};



class Solution {
public:
    int maxSizeSlices(vector<int>& slices) {
        unordered_map<int, unordered_map<int, unordered_map<int, int>>>mp;
        return dp(0,slices.size()-1, slices.size()/3, slices, mp, 1);
    }
    
    int dp(int i , int j, int k, const vector<int>&slices, unordered_map<int, unordered_map<int, unordered_map<int, int>>>&mp, int lastOne = 0){
        if(k == 1)
            return *max_element(slices.begin()+i, slices.begin()+j+1);
        if(j-i + 1 < 2*k - 1)
            return numeric_limits<int>::min();
        if(mp.count(i) && mp[i].count(j) && mp[i][j].count(k))
            return mp[i][j][k];
        return mp[i][j][k] = max(slices[j] + dp(i + lastOne, j-2, k-1, slices, mp), dp(i, j-1, k, slices,mp));;
    }
};